fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java

```diff
@@ -0,0 +1,292 @@
+                                    .filter(Optional::isPresent)
+                                    .map(Optional::get)
+                                    .findFirst())
+                    .map(Supplier::get)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .findFirst();
+        }
+        if (input.isNullType()) {
+            // Cast null type (usually from null literals) into target types
+            returnType = expected.defaultConcreteType();
+        } else if (input.isNumericType() && expected instanceof DecimalType) {
+            // If input is a numeric type but not decimal, and we expect a decimal type,
+            // cast the input to decimal.
+            returnType = DecimalType.forType(input);
+        } else if (input.isNumericType() && expected instanceof NumericType) {
+            // For any other numeric types, implicitly cast to each other, e.g. bigint -> int, int -> bigint
+            returnType = (DataType) expected;
+        } else if (input.isStringType()) {
+            if (expected instanceof DecimalType) {
+                returnType = DecimalType.SYSTEM_DEFAULT;
+            } else if (expected instanceof NumericType) {
+                returnType = expected.defaultConcreteType();
+            }
+        } else if (input.isPrimitive() && !input.isStringType() && expected instanceof CharacterType) {
+            returnType = StringType.INSTANCE;
+        }
+
+        // could not do implicit cast, just return null. Throw exception in check analysis.
+        return Optional.ofNullable(returnType);
+    }
+
+    /**
+     * return ture if two type could do type coercion.
+     */
+    public static boolean canHandleTypeCoercion(DataType leftType, DataType rightType) {
+        if (leftType instanceof DecimalType && rightType instanceof NullType) {
+            return true;
+        }
+        if (leftType instanceof NullType && rightType instanceof DecimalType) {
+            return true;
+        }
+        if (!(leftType instanceof DecimalType) && !(rightType instanceof DecimalType) && !leftType.equals(rightType)) {
+            return true;
+        }
+        return false;
+    }
+
+
+
+    /**
+     * find the tightest common type for two type
+     */
+    @Developing
+    public static Optional<DataType> findTightestCommonType(DataType left, DataType right) {
+        // TODO: compatible with origin planner and BE
+        // TODO: when add new type, add it to here
+        DataType tightestCommonType = null;
+        if (left.equals(right)) {
+            tightestCommonType = left;
+        } else if (left instanceof NullType) {
+            tightestCommonType = right;
+        } else if (right instanceof NullType) {
+            tightestCommonType = left;
+        } else if (left instanceof IntegralType && right instanceof DecimalType
+                && ((DecimalType) right).isWiderThan(left)) {
```

Comment:
do we missed the case:
left is integral and right is decimal but  (NOT right.isWiderThan(left))

Reply:
it will be processed by findWiderTypeForDecimal